transport/src/main/java/io/netty/channel/embedded/EmbeddedChannel.java

```diff
@@ -875,4 +923,30 @@ protected void onUnhandledInboundMessage(ChannelHandlerContext ctx, Object msg)
             handleInboundMessage(msg);
         }
     }
+
+    private static final class EmbeddedChannelPromise extends DefaultChannelPromise {
+        EmbeddedChannelPromise(Channel channel) {
+            super(channel);
+        }
+
+        EmbeddedChannelPromise(Channel channel, EventExecutor executor) {
+            super(channel, executor);
+        }
+
+        @Override
+        protected boolean notifyWithExecutor() {
+            return false;
+        }
+    }
+
+    private static final class EmbeddedChannelProgressivePromise extends DefaultChannelProgressivePromise {
+        EmbeddedChannelProgressivePromise(Channel channel, EventExecutor executor) {
+            super(channel, executor);
+        }
+
+        @Override
+        protected boolean notifyWithExecutor() {
+            return false;
```

Comment:
yeah... as it basically would mean that you need to call `runPendingTasks()` every time a future is used.

Reply:
I was saying that _most_ times a future a used runPendingTasks() is already run. But I agree that all the futures returned by EmbeddedChannel (itself, not from ctx) would cause trouble, since the test would add a listener after embedded channel returns.

This makes embedded channel even more unhelpful to use as a test, since the app will still see reentrancy.

What if we made EmbeddedEventLoop smarter instead? Let's say we add an option: "eager execute" (for backward compatibility; or we just change the behavior). When `EmbeddedEventLoop.execute()` is called it will call runPendingTasks() immediately, unless it is already running. `inEventLoop()` only returns true while it is executing. The various calls on the channel would enqueue onto the event loop (via ChannelHandlerContext) when run from the test, but would be immediate from handlers (just like in "real life"). Any `addListener()` calls after the channel returns to the test (say, after a `channel.write()`) will execute immediately. This would mean calling runPendingTasks() would be unnecessary, except for scheduled tasks.

Reply:
@ejona86 let me do some testing... Will come back to you.

Reply:
@ejona86 I did a bit of testing and I think this is kind of tricky as a lot of use-cases asume that `inEventLoop` will just return true always...  Also after thinking more about this I wonder if we even need it at all after https://github.com/netty/netty/pull/9358 . WDYT ?

Reply:
@ejona86 thanks for being so persistent here buddy :) After thinking more about all of this I think you are right and addressing this will even make our unit-tests less dependent on the internals of `EmbeddedChannel`. Be prepared for an updated PR soon. Just working through the last things here...